File: 2181.merge-nodes-in-between-zeros.py
#
# @lc app=leetcode id=2181 lang=python3
#
# [2181] Merge Nodes in Between Zeros
#

# @lc code=start
# Definition for singly-linked list.
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # sum = 0
        # cur = head
        # ans = ListNode(0)
        # tmp = ans
        # while cur:
        #     while cur.val != 0:
        #         sum += cur.val
        #         cur = cur.next
        #     if sum != 0:
        #         tmp.next = ListNode(sum)
        #         tmp = tmp.next
        #     cur = cur.next
        #     sum = 0
        # return ans.next
        sum = 0
        head2 = head
        cur = head.next
        while cur.next: 
            if cur.val == 0:
                head2 = head2.next
                head2.val = 0
            else:
                head2.val += cur.val
            cur = cur.next
        head2.next = None
        return head


        

# @lc code=end

File: 2181.merge-nodes-in-between-zeros.java
/*
 * @lc app=leetcode id=2181 lang=java
 *
 * [2181] Merge Nodes in Between Zeros
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeNodes(ListNode head) {
        ListNode cur = head.next;
        ListNode concate = head;
        while(cur.next != null){
            if(cur.val != 0){
                concate.val += cur.val;
            }
            else{
                concate = concate.next;
                concate.val = 0;
            }
            cur = cur.next;
        }
        concate.next = null;
        return head;
    }

}
// @lc code=end

